### PR TITLE
Add LargeArrayBuilder type to enscapulate ToArray logic

### PR DESCRIPTION
**Pull request by jamesqo**
The `EnumerableHelpers.ToArray` logic I introduced in #11208, while better in terms of memory consumption, employs a very imperative code style. I refactored the core logic into a new struct called `LargeArrayBuilder` so now the method looks like

``` cs
// Check for ICollection

var builder = new LargeArrayBuilder(initialize: true);
builder.AddRange(source);
return builder.ToArray();
```

As a result of this refactoring, we can use the buffering strategy employed by `ToArray` in other places where we were previously using `EnumerableHelpers.ToArray(this)` (such as `Select.ToArray` or `Concat.ToArray`) and simultaneously avoid some unnecessary indirection/field stores. This results in substantial speedup.

Perf testing and analysis: [test code](https://github.com/jamesqo/Dotnet/blob/4f08d2a4ce3c2aedf9bafec1516ffe3d86c22e81/Program.cs), [data](https://gist.github.com/jamesqo/2c7c07ccf64a52f0d52c1bd8902e407d).

The following patterns all see substantial speedups (where `e` is a lazy enumerable):
- `e.Concat.ToArray`
- `e.Append.ToArray`
- `e.Append.Append.ToArray`
- `e.Select.ToArray`
- `e.OrderBy.Select.ToArray` (OrderBy returns an IPartition that sometimes can't get its count cheaply)

Once this is merged I will change #12703 to make `Where` use it too, so all `Where.ToArray` calls should be substantially faster.

cc @stephentoub @jkotas @VSadov @JonHanna

### CODE REVIEW

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        {
+            // - On the first few adds, simply resize _first.
+            // - When we pass ResizeLimit, allocate ResizeLimit elements for _current
+            //   and start reading into _current. Set _index to 0.
+            // - When _current runs out of space, add it to _buffers and repeat the
+            //   above step, except with _current.Length * 2.
+
+            Debug.Assert(_index == _current.Length, $"{nameof(AllocateBuffer)} was called, but there's more space.");
+
+            T[] result;
```
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: self-note: Remove this variable

**Comment by karelz** [comment]
cc: @Priya91 @ianhays

**Comment by jamesqo** [comment]
@karelz, FYI I'd say this PR is more related to System.Linq than System.Collections.

**Comment by jamesqo** [comment]
@karelz, FYI I'd say this PR is more related to System.Linq than System.Collections. Would be nice if you could relabel / change assignees.

**Comment by karelz** [comment]
I see the code changes in both -- Collections and Linq. I will leave area owners/experts to decide who is in best position to do the code review.

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+            {
+                AddRange(enumerator);
+            }
+        }
+
+        /// <summary>
+        /// Adds a range of items to this builder.
+        /// </summary>
+        /// <param name="enumerator">The enumerator to add.</param>
+        public void AddRange(IEnumerator<T> enumerator)
```
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: TODO: It may be worth eliminating this and just writing the whole thing inline in `AddRange(IEnumerable)`. The one place where it's used can be eliminated by adding some extra methods to `SingleLinkedNode`.

**Comment by jamesqo** [comment]
@stephentoub Do you think it would be possible for you to review this when you have time? Thanks.

**Comment by karelz** [comment]
@ianhays @Priya91 @VSadov can you please review the code?

**Comment by karelz** [comment]
@ianhays @Priya91 @VSadov @OmarTawfik ping? Who's the best person to review the change?

File: src/Common/tests/Tests/System/Collections/Generic/ArrayBuilderTests.cs [hunk 3]
```diff
@@ -13,6 +13,22 @@ public interface IGenerator<out T>
         T Generate(int seed);
     }
 
+    public static class GeneratorExtensions
+    {
+        public static IEnumerable<T> GenerateSequence<T>(this IGenerator<T> generator, int count)
```
Comment (stephentoub) [inline src/Common/tests/Tests/System/Collections/Generic/ArrayBuilderTests.cs]: Why did you rename this from GenerateEnumerable to GenerateSequence?

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 4]
```diff
@@ -166,12 +166,33 @@ public override AppendPrependIterator<TSource> Prepend(TSource item)
...
+                {
+                    builder.SlowAdd(_item);
+                }
+
+                builder.AddRange(_source);
+
+                if (_appending)
+                {
+                    builder.SlowAdd(_item);
+                }
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Would this not be better / simpler as:
```C#
if (_appending)
{
    builder.AddRange(_source);
    builder.SlowAdd(_item);
}
else
{
    builder.SlowAdd(_item);
    builder.AddRange(_source);
}
```
?

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 5]
```diff
@@ -272,6 +293,11 @@ private SingleLinkedNode(SingleLinkedNode<TSource> linked, TSource item)
 
             public IEnumerator<TSource> GetEnumerator()
             {
+                return ToArray().AsEnumerable().GetEnumerator();
+            }
+
+            public TSource[] ToArray()
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Why separate this out into its own (public) method?  Where else is it being used?

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 6]
```diff
@@ -352,12 +378,36 @@ public override AppendPrependIterator<TSource> Prepend(TSource item)
...
+                for (SingleLinkedNode<TSource> node = _prepended; node != null; node = node.Linked)
+                {
+                    builder.Add(node.Item);
+                }
+
+                builder.AddRange(_source);
+
+                if (_appended != null)
+                {
+                    foreach (TSource item in _appended.ToArray())
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Why are we creating this intermediate array?

File: src/System.Linq/src/System/Linq/Select.cs [hunk 7]
```diff
@@ -625,13 +635,25 @@ public TResult TryGetLast(out bool found)
                 return sourceFound ? _selector(input) : default(TResult);
             }
 
+            private TResult[] LazyToArray()
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Select.cs]: Why separate this out?  The other cases in ToArray aren't in their own methods, seems like this should just be inline with the others.

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+        private ArrayBuilder<T[]> _buffers; // After ResizeLimit * 2, we store previous buffers we've filled out here.
+        private T[] _current;               // Current buffer we're reading into. Is _first <= ResizeLimit.
+        private int _index;                 // Index into the current buffer.
+        private int _count;                 // Count of all of the items in this builder.
+
+        /// <summary>
+        /// Constructs a new builder.
+        /// </summary>
+        /// <param name="initialize">Pass <c>true</c>.</param>
+        public LargeArrayBuilder(bool initialize) : this()
```
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: Instead of the ```: this()```, how about just explicitly initializing the remaining fields (_buffers, _index, _count)?

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+        /// <summary>
+        /// Adds an item to this builder.
+        /// </summary>
+        /// <param name="item">The item to add.</param>
+        /// <remarks>
+        /// Use <see cref="Add"/> if adding to the builder is a bottleneck for your use case.
+        /// Otherwise, use <see cref="SlowAdd"/>.
+        /// </remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Add(T item)
```
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: > I found out this wasn't getting inlined, so I tagged it with AggressiveInlining

Why was it important that it get inlined?  Do you have a benchmark that shows a measurable improvement in key scenarios from that happening?  Otherwise I'd rather let the JIT do its thing and not complicate the API here with two Add methods that are identical other than in inlining.

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+                {
+                    _buffers.Add(_current);
+                    nextCapacity = _current.Length * 2;
+                }
+
+                _current = new T[nextCapacity];
+                _index = 0;
+            }
+            
+            return _current;
```
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: Why have AllocateBuffer return _current?  We don't always use the result... why not just have the call sites that need _current access _current?

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+
+                // Continuously read in items from the enumerator, updating _count
+                // and _index when we run out of space.
+
+                while (enumerator.MoveNext())
+                {
+                    if (index == destination.Length)
+                    {
+                        // No more space in this buffer. Resize.
+                        _count += index - _index;
```
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: Can this overflow?

**Comment by stephentoub** [comment]
Thanks, @jamesqo.  I left some comments to be addressed, but overall it looks good.

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 9 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: @stephentoub, will post when I get a chance.

File: src/Common/tests/Tests/System/Collections/Generic/ArrayBuilderTests.cs (hunk 3 above)
Comment (jamesqo) [inline src/Common/tests/Tests/System/Collections/Generic/ArrayBuilderTests.cs]: @stephentoub, can rename back.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 4 above)
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: @stephentoub, this is the same pattern that is used in other places throughout the file, e.g. [here](https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/AppendPrepend.cs#L211).

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 12]
```diff
@@ -272,6 +293,11 @@ private SingleLinkedNode(SingleLinkedNode<TSource> linked, TSource item)
 
             public IEnumerator<TSource> GetEnumerator()
             {
+                return ToArray().AsEnumerable().GetEnumerator();
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: I personally like `AsEnumerable` better, but sure.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 6 above)
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: @stephentoub `LargeArrayBuilder` does not have an indexer, since it cannot be implemented efficiently. (I did it before using some bitshifting tricks/log2, but it was too slow.) In addition, we don't know if there's even enough room in the buffers to hold the appended items-- for example, we could have added 40 items, meaning there's 64 slots, and we need to append 25 items.

File: src/System.Linq/src/System/Linq/Select.cs (hunk 7 above)
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Select.cs]: @stephentoub, `ToArray` is generic so it can be regenerated by the JIT. The logic in `LazyToArray` seemed big enough that it seemed worth separating out into a new method, to avoid penalizing code that doesn't go down this branch. I'll post the code size of this method when I get time.

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 8 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: @stephentoub, is explicit initialization faster than using the default struct constructor?

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 10 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: Ok

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 11 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: Not before we get an OOME.

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 9 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: Yes, removing the AggressiveInlining indeed does result in a perf regression. I wrote [this app](https://gist.github.com/jamesqo/6e05ecf2ea0da3569a86f5f4fcb1326d) to measure the throughput with/without the AggressiveInlining; removing the attribute results in a significant (~30%) slowdown.

File: src/System.Linq/src/System/Linq/Select.cs (hunk 7 above)
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Select.cs]: Just checked. `LazyToArray` comes in at an extra 306 bytes of code: https://gist.github.com/jamesqo/6b5f2cc27ca36fd3f488c7be57a87cb4 It would probably be wise to avoid generating all that unnecessary code if it's never called.

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 8 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: I just checked. If we initialize members explicitly then the constructor doesn't get inlined:

```asm
G_M36021_IG02:
       488D4DC8             lea      rcx, bword ptr [rbp-38H]
       BA01000000           mov      edx, 1
       E88CFAFFFF           call     LargeArrayBuilder`1:.ctor(bool):this
```

While this isn't the bottleneck, I think using `: this()` looks more readable anyway.

**Comment by jamesqo** [comment]
@stephentoub, thanks for reviewing. I've responded to all of your comments.

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 8 above)
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: > I think using : this() looks more readable anyway.

I personally disagree.  To me this looks like we're initializing some of the fields multiple times, as delegating to "this()" needs to ensure that all of the fields are initialized, and then we subsequently overwrite some of them (it looks to me the same as having a bunch of fields in a class that are explicitly initialized to default(T) at the field declaration site and then subsequently overwritten in an explicit ctor). We also don't use this pattern elsewhere in corefx, which makes this stand out to me like a sore thumb.

If there's actually an inlining difference and it matters, that's something that should be addressed in the JIT.
cc: @AndyAyersMS

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 9 above)
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: > removing the attribute results in a significant (~30%) slowdow

You're comparing release builds, and the only difference between them is whether Add has AggressiveInlining on it?  I just tried with your PR (with and without that one line with AggressiveInlining), and while I do see an improvement that comes from the PR in general, I don't see any such difference between whether Add has AggressiveInlining on it.

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 8 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: @stephentoub 

> it looks to me the same as having a bunch of fields in a class that are explicitly initialized to default(T) at the field declaration site and then subsequently overwritten in an explicit ctor

If you think about it, that can be faster since the zeroing can be done in bulk vs. initializing fields 1 at a time. For example, on my machine the jit generates code to use `ymm` registers:

```asm
       C4E17957C0           vxorpd   ymm0, ymm0
       C4E17A7F02           vmovdqu  qword ptr [rdx], ymm0
       C4E17A7F4210         vmovdqu  qword ptr [rdx+16], ymm0
```

> If there's actually an inlining difference and it matters, that's something that should be addressed in the JIT.

Considering things from the viewpoint of the JIT, isn't the constructor just a bunch of arbitrary code? So then when it looks at the constructor and sees that it has a huge body, it might not decide to inline even though all of the code is just default-initializing the struct. It would seem that if the JIT were to try to inline this, it would waste time for other cases where that arbitrary code isn't all just default-initialization.

I'm not really familiar with how the JIT works, but just a guess... cc @mikedn, @benaadams

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 9 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: @stephentoub 

> You're comparing release builds, and the only difference between them is whether Add has AggressiveInlining on it?

Yes. What test app are you using?

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 8 above)
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: > If you think about it, that can be faster since the zeroing can be done in bulk vs. initializing fields 1 at a time

I'm not talking about what actually happens in the implementation; you were talking about readability, and part of readability is being able to reason about what's happening, so I was sharing what immediately comes to my mind when I see that code, regardless of what the compiler does, and hence why for me it harms readability, along with the fact that readability is helped by seeing common patterns and harmed by seeing jarring patterns, and to me, this use of ": this()" in a struct is jarring.

A compiler can choose to optimize certain patterns, so the compiler could choose to generate the exact same code for:
```C#
public SomeStruct()
{
    _field1 = 0;
    _field2 = null;
    _field3 = Calculate();
}
```
as it does for:
```
public SomeStruct() : this()
{
    _field3 = Calculate();
}
```

Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: > along with the fact that readability is helped by seeing common patterns and harmed by seeing jarring patterns, and to me, this use of ": this()" in a struct is jarring.

OK, I can remove the explicit `: this()` call then. Sorry.

> A compiler can choose to optimize certain patterns, so the compiler could choose to generate the exact same code for:

Yes. I was trying to say that with the first version the JIT would (probably) have to inspect more bytecode to see that `0` was being assigned to `_field1`, and `null` was being assigned to `_field2`, before making that optimization (which would presumably come at a cost since all of this is done at runtime).

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 11 above)
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: > Not before we get an OOME.

Won't it overflow when destination.Length == 0x40000000?

Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: @stephentoub, good point. In that case, however, `AllocateBuffer` will attempt to allocate an array of size 0x40000000 * 2, which will raise an OverflowException anyways. Also, if we've reached that point then that would mean we've added 0x80000000 items, meaning `ToArray` couldn't possibly work. Idk if it's worth adding an extra `checked` block which has no effect, and protects against a case which is already broken.

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 9 above)
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: > Yes. What test app are you using to benchmark?

Same as you.  I've tried a few other things, though, and I am able to construct a simple console app that sees about a ~15% improvement from the AggressiveInlining, so while not 30%, it does seem worthwhile to keep it for now.

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 11 above)
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: > however, AllocateBuffer will attempt to allocate an array of size 0x40000000 * 2, which will raise an OverflowException anyways

Will it?  I've not tried it, but it looks like it would actually see _count as negative, which is less than ResizeLimit, so it would enter the "we're still less than ResizeLimit" block.  Maybe those checks need to be made robust.

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 9 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: @stephentoub, good. My percentages may have been skewed a bit since I was comparing the smaller times in relation to the larger times instead of vice versa (1100 is a 20% regression from 900, 900 is an 18% improvement from 1100) so I will have to be careful about that in the future.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Ok.  For consistency it's worth keeping it like this for now, but I think it'd also be worth following up to see whether it'd be worth changing all of them to what, at least to me, looks like a simpler and more understandable pattern.  I don't know what the JIT'd code looks like it, but it's possible there's also a perf benefit to one or the other, either in number of actual branches or in IL size or something like that.

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 11 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: Ah, very nice catch. It looks like we will multiply `_count` by 2 again, and 0x80000000 * 2 is 0, so then we allocate a 0-length array and set it to `_first` / `_current` instead of raising an exception. Since it doesn't add any additional cost, I can change the check to be `if ((uint)_count < (uint)ResizeLimit)` which should get us down the path which throws for overflow.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 6 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Was such an array getting allocated before?  I'm struggling to see this as a pure win if we're introducing additional intermediate ToArrays like this.  If it was already there before, then ok, though it would be nice subsequently to find a way to avoid it.

Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: @stephentoub [Yes.](https://github.com/jamesqo/corefx/blob/6c7675111fb3854b0e5535779a1dd04b8f4bd833/src/System.Linq/src/System/Linq/AppendPrepend.cs#L360) The logic in this method is basically a copy verbatim of what `MoveNext` currently does.

File: src/System.Linq/src/System/Linq/Select.cs (hunk 7 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Select.cs]: > It would probably be wise to avoid generating all that unnecessary code if it's never called.

Why is that the case for "case -1" but not the case for "default" in ToArray?  My point is that we should be consistent here.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 6 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: > Yes. The logic in this method is basically a copy verbatim of what MoveNext currently does.

Ugh. Ok. Thanks.

File: src/System.Linq/src/System/Linq/Select.cs (hunk 7 above)
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Select.cs]: @stephentoub, I separated the `default:` case into a new `PreallocatingToArray` method.

**Comment by jamesqo** [comment]
Test Innerloop OSX Debug Build and Test
Innerloop Linux ARM Emulator SoftFP Debug Build and Test
Innerloop Linux ARM Emulator SoftFP Release Build and Test

**Comment by jamesqo** [comment]
Test Innerloop Linux ARM Emulator SoftFP Debug Build and Test
Test Innerloop Linux ARM Emulator SoftFP Release Build and Test

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 9 above)
Comment (AndyAyersMS) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: Sorry, @jamesqo just saw this now.... did you ever drill deeper into the codegen with and without forced inlining?

Comment (AndyAyersMS) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: Ok, thanks. Not obvious from just this where the win came from.  I'd need to go back and look at what Add compiles to when it's not inlined.
